File: pr2 full/Задание 2 Практика Go/5.go
package main

import "fmt"

func runTask5() {
	p := 23
	g := 5
	a := 6
	b := 15
	e := 9
	message := "HELLO"

	A := powerMod(g, a, p)
	B := powerMod(g, b, p)

	fakeForBob := powerMod(g, e, p)
	fakeForAlice := powerMod(g, e, p)

	keyAliceMallory := powerMod(fakeForAlice, a, p)
	keyBobMallory := powerMod(fakeForBob, b, p)
	keyMalloryWithAlice := powerMod(A, e, p)
	keyMalloryWithBob := powerMod(B, e, p)

	encryptedByAlice := encryptText(message, keyAliceMallory)
	readByMallory := decryptText(encryptedByAlice, keyMalloryWithAlice)
	changedMessage := readByMallory + "!"
	encryptedForBob := encryptText(changedMessage, keyMalloryWithBob)
	readByBob := decryptText(encryptedForBob, keyBobMallory)

	fmt.Println("\n--- Задание 5* ---")
	fmt.Println("Эмуляция атаки посередине на обмен Диффи-Хеллмана")
	fmt.Println("p =", p, ", g =", g)
	fmt.Println("Алиса отправляет A =", A, ", Боб отправляет B =", B)
	fmt.Println("Злоумышленник подменяет оба значения на E = g^e mod p =", fakeForBob)
	fmt.Println("Ключ Алиса-злоумышленник:", keyAliceMallory)
	fmt.Println("Ключ злоумышленник-Алиса:", keyMalloryWithAlice)
	fmt.Println("Ключ Боб-злоумышленник:", keyBobMallory)
	fmt.Println("Ключ злоумышленник-Боб:", keyMalloryWithBob)
	fmt.Println("Исходное сообщение Алисы:", message)
	fmt.Println("Злоумышленник прочитал:", readByMallory)
	fmt.Println("Боб получил после подмены:", readByBob)
	fmt.Println("\nКраткое сообщение по теме находится в файле 5_сообщение.txt.")

	waitEnter()
}

File: pr2 full/5.cpp
#include <iostream>
#include <string>

#include "func.h"

using namespace std;

void runTask5() {
    int p = 23;
    int g = 5;
    int a = 6;
    int b = 15;
    int e = 9;
    string message = "HELLO";

    int A = powerMod(g, a, p);
    int B = powerMod(g, b, p);

    int fakeForBob = powerMod(g, e, p);
    int fakeForAlice = powerMod(g, e, p);

    int keyAliceMallory = powerMod(fakeForAlice, a, p);
    int keyBobMallory = powerMod(fakeForBob, b, p);
    int keyMalloryWithAlice = powerMod(A, e, p);
    int keyMalloryWithBob = powerMod(B, e, p);

    string encryptedByAlice = encryptText(message, keyAliceMallory);
    string readByMallory = decryptText(encryptedByAlice, keyMalloryWithAlice);
    string changedMessage = readByMallory + "!";
    string encryptedForBob = encryptText(changedMessage, keyMalloryWithBob);
    string readByBob = decryptText(encryptedForBob, keyBobMallory);

    cout << "\n--- Задание 5* ---" << endl;
    cout << "Эмуляция атаки посередине на обмен Диффи-Хеллмана" << endl;
    cout << "p = " << p << ", g = " << g << endl;
    cout << "Алиса отправляет A = " << A << ", Боб отправляет B = " << B << endl;
    cout << "Злоумышленник подменяет оба значения на E = g^e mod p = " << fakeForBob << endl;
    cout << "Ключ Алиса-злоумышленник: " << keyAliceMallory << endl;
    cout << "Ключ злоумышленник-Алиса: " << keyMalloryWithAlice << endl;
    cout << "Ключ Боб-злоумышленник: " << keyBobMallory << endl;
    cout << "Ключ злоумышленник-Боб: " << keyMalloryWithBob << endl;
    cout << "Исходное сообщение Алисы: " << message << endl;
    cout << "Злоумышленник прочитал: " << readByMallory << endl;
    cout << "Боб получил после подмены: " << readByBob << endl;
    cout << "\nКраткое сообщение по теме находится в файле 5_сообщение.txt." << endl;

    waitEnter();
}
